hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java

```diff
@@ -109,22 +126,116 @@ public class RecordCacheImpl implements HederaRecordCache {
      */
     private final DeduplicationCache deduplicationCache;
     /**
-     * A map of transaction IDs to the histories of all transactions that came to
-     * consensus with that ID, or their child
-     * transactions. This data structure is rebuilt during reconnect or restart. Using a non-deterministic, map is
-     * perfectly acceptable, as the order of these histories is not important.
+     * A map of transaction IDs to the sources of records for those transaction ids and their children.
      */
-    private final Map<TransactionID, History> histories;
+    private final Map<TransactionID, HistorySource> historySources = new ConcurrentHashMap<>();
     /**
      * A secondary index that maps from the AccountID of the payer account to a set of transaction IDs that were
      * submitted by this payer. This is only needed for answering queries. Ideally such queries would exist on the
      * mirror node instead. The answer to this query will include child records that were created as a consequence
      * of the original user transaction, but not any preceding records triggered by it.
      */
-    private final Map<AccountID, Set<TransactionID>> payerToTransactionIndex = new ConcurrentHashMap<>();
-
+    private final Map<AccountID, Set<TransactionID>> payerTxnIds = new ConcurrentHashMap<>();
+    /**
+     * The list of transaction receipts for the current round.
+     */
     private final List<TransactionReceiptEntry> transactionReceipts = new ArrayList<>();
 
+    /**
+     * Contains history of transactions submitted with the same "base" {@link TransactionID};
+     * i.e., with the same payer and valid start time.
+     * <p>
+     * This history has two parts:
+     * <ol>
+     *     <li>A {@code nodeIds} set with all the node ids that have submitted a properly
+     *     screened transaction with the scoped base {@link TransactionID}; this is used to
+     *     classify duplicate transactions.</li>
+     *     <li>A {@code recordSources} list with all the sources of records for the relevant
+     *     base {@link TransactionID}. This is used to construct {@link TransactionRecord}
+     *     records for answering queries.</li>
+     * </ol>
+     *
+     * @param nodeIds The set of node ids that have submitted a properly screened transaction
+     * @param recordSources The sources of records for the relevant base {@link TransactionID}
+     */
+    private record HistorySource(@NonNull Set<Long> nodeIds, @NonNull List<RecordSource> recordSources)
```

Comment:
Since we maintain most of the objects per `TransactionID`, is it difficult to have `RecordSource` maintain per `TransactionID` ?

Reply:
There is just performance overhead I think, not _that_ much difficulty. We should definitely keep thinking about this...